library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs

```diff
@@ -1,65 +1,216 @@
+-- The function returns the name of the create function.
+mkLam ::
+  (LLVM.MonadIRBuilder m, LLVM.MonadModuleBuilder m) =>
+  -- | Environment of Juvix variables to LLVM function arguments.
+  Env ->
+  -- | The type of the lambda abstraction.
+  ErasedAnn.Type PrimTy ->
+  -- | The body of the lambda abstraction.
+  ErasedAnn.AnnTerm PrimTy RawPrimVal ->
+  -- | List of parameter names.
+  [NameSymbol.T] ->
+  -- | List of captures variables (free variables for the body).
+  [NameSymbol.T] ->
+  m LLVM.Name
+mkLam env ty body args capt = do
+  funname <- LLVM.freshName "lam"
+  let returnTy = typeToLLVM $ ErasedAnn.type' body
+  params <- mapM mkParam (zip args (functionTy ty))
+  LLVM.function funname params returnTy $ \refs -> do
+    let env' = env `Map.union` Map.fromList (zip args refs)
+    body' <- compileTerm env' body
+    LLVM.ret body'
+  return funname
+  where
+    -- Given a Juvix name and type, construct an LLVM function parameter
+    -- definition.
+    mkParam ::
+      LLVM.MonadIRBuilder m =>
+      (NameSymbol.T, ErasedAnn.Type PrimTy) ->
+      m (LLVM.Type, LLVM.ParameterName)
+    mkParam (name, ty) = do
+      let ty' = typeToLLVM ty
+          name' = unintern $ NameSymbol.toSymbol name
+      return $ (ty', S.fromString $ "arg" <> name')
 
+-- | The function assumes the arguments passed are the arguments of an
+-- application.
 mkApp ::
-  ErasedAnn.Term PrimTy RawPrimVal ->
+  (LLVM.MonadIRBuilder m, LLVM.MonadModuleBuilder m) =>
+  -- | Environment of Juvix variables to LLVM function arguments.
+  Env ->
+  -- | The function term of an application.
+  ErasedAnn.AnnTerm PrimTy RawPrimVal ->
+  -- | The type of the application.
   ErasedAnn.Type PrimTy ->
+  -- | The arguments to the application.
   [ErasedAnn.AnnTerm PrimTy RawPrimVal] ->
-  LLVM.IRBuilderT LLVM.ModuleBuilder LLVM.Operand
-mkApp f (ErasedAnn.PrimTy ty) xs = case f of
-  ErasedAnn.Prim prim -> applyPrim prim ty xs
+  m LLVM.Operand
+mkApp env f@(ErasedAnn.Ann {ErasedAnn.term, ErasedAnn.type'}) _ xs =
+  case term of
+    ErasedAnn.LamM {ErasedAnn.body, ErasedAnn.arguments, ErasedAnn.capture} -> do
+      funname <- mkLam env type' body arguments capture
+      xs' <- mapM (compileTerm env) xs
+      let xs'args = zip xs' (repeat [])
+      LLVM.call (globalRef (typeToLLVM type') funname) xs'args
+    ErasedAnn.Prim prim -> applyPrim env prim xs
+    ErasedAnn.Var v -> do
+      f' <- compileTerm env f
+      xs' <- mapM (compileTerm env) xs
+      let xs'args = zip xs' (repeat []) -- Do not pass attributes to the args.
+      LLVM.call f' xs'args
+
+-- | Write LLVM code for a primitive.
+-- TODO: implement other primitives.
+mkPrim ::
+  LLVM.MonadIRBuilder m =>
+  -- | Term that contains the primitive.
+  ErasedAnn.Term PrimTy RawPrimVal ->
+  -- | Type of the primitive.
+  ErasedAnn.Type PrimTy ->
+  m LLVM.Operand
+mkPrim (ErasedAnn.Prim prim) ty = case prim of
+  LitInt i -> case ty of
```

Comment:
seems like we need more cases or some kind of error case?

Reply:
Currently the set of supported primitives is extremely limited and should definitely be extended either right before or right after the release.

I can add an error here, but there no other primitives that could be used. Having a fall-through case keeps GHC from spitting compiler warnings as well, but we actually want them for functions like this, to remind us of the intotality.

Reply:
Hmm sure, I'm not against this but I don't think it's very well standardised across the codebase.

Reply:
No it isn't really standardised across the codebase. It doesn't help that GHC gives a rather generic `Non-exhaustive patterns in case` instead of telling the exact case that didn't have a match, because that would be the nicest solution as it doesn't require us to do anything.